### DISCUSSION TITLE
Add support for topology-aware placement of EC shards

### DISCUSSION THREAD

**Discussion by ghost** [seaweedfs/seaweedfs#6179] [Ideas]
Seaweed is currently not taking cluster topology into consideration when EC shards are created/spread, which can lead to sub-optimal allocations - including potential data loss scenarios if multiple EC shards are created on a single server.

See https://github.com/seaweedfs/seaweedfs/issues/2705 for a bug report where this issue is highlighted.

**Comment by ghost** [comment]
I'm currently working on a couple MR to contribute this feature, if Seaweed devs are interested 🙂

The first step would consist in unifying existing EC logic (for `ec.balance` and `ec.encode`) into a common module, then rework this logic so EC takes topology into consideration when shards are created.

**Comment by jimmyspets** [comment]
Does this meen that the EC algorithm would risk placing to many shards on the same physical server if I use the example from the production setup of serving each physical disk on a separate port?
Would that be counted as three different servers?

**Reply by ghost** [reply]
> Does this meen that the EC algorithm would risk placing to many shards on the same physical server if I use the example from the production setup of serving each physical disk on a separate port? Would that be counted as three different servers?

Note that a setup with three volume servers cannot really do EC safely :disappointed: 

Seaweed uses RS(10,4) for erasure coding; having three volume servers means that inevitably at least one of them would store 1+ EC volumes, and data can no longer be read if that server is down.

**Comment by ghost** [comment]
`ec.balance` is now topology-aware. The command takes a parameter to configure replica placement settings for EC volumes; if left unspecified, the master's default replica placement is used instead. Volume placement will then obey this setting.

`ec.encode` will follow soon.

Moving forward it would be nice for seaweed to support proper EC shard placement strategy settings, but this approach should suffice in the meantime.

**Comment by ghost** [comment]
`ec.encode` is now topology-aware: https://github.com/seaweedfs/seaweedfs/pull/6339

**Comment by bradmurray** [comment]
I assume this will be releases post 3.80

**Reply by ghost** [reply]
> I assume this will be releases post 3.80

That's up to @chrislusf but, yes, i presume the next stable release should contain all of these changes.

**Comment by ghost** [comment]
Documentation is now updated: https://github.com/seaweedfs/seaweedfs/wiki/Erasure-Coding-for-warm-storage#ec-data-balancing

**Reply by ghost** [reply]
...aaaaaaaaaand with https://github.com/seaweedfs/seaweedfs/pull/6377 being merged, this should be ready to go as it is :partying_face:
